vortex-dtype/src/arrow.rs

```diff
@@ -150,73 +168,148 @@ impl DType {
                         field_name.to_string(),
-                        field_dt.to_arrow_dtype()?,
+                        field_dt.to_arrow_field()?.data_type().clone(),
                         field_dt.is_nullable(),
                     )));
                 }
 
-                DataType::Struct(Fields::from(fields))
+                Field::new(
+                    "_default",
+                    DataType::Struct(Fields::from(fields)),
+                    (*n).into(),
+                )
             }
             // There are four kinds of lists: List (32-bit offsets), Large List (64-bit), List View
             // (32-bit), Large List View (64-bit). We cannot both guarantee zero-copy and commit to an
             // Arrow dtype because we do not how large our offsets are.
-            DType::List(l, _) => DataType::List(FieldRef::new(Field::new_list_field(
-                l.to_arrow_dtype()?,
-                l.nullability().into(),
-            ))),
+            DType::List(elem_type, n) => Field::new(
+                "_default",
+                DataType::List(FieldRef::new(Field::new_list_field(
+                    elem_type.to_arrow_field()?.data_type().clone(),
+                    elem_type.nullability().into(),
+                ))),
+                (*n).into(),
+            ),
             DType::Extension(ext_dtype) => {
                 // Try and match against the known extension DTypes.
                 if is_temporal_ext_type(ext_dtype.id()) {
-                    make_arrow_temporal_dtype(ext_dtype)
+                    Field::new(
+                        "_default",
+                        make_arrow_temporal_dtype(ext_dtype),
+                        ext_dtype.storage_dtype().is_nullable(),
+                    )
                 } else {
-                    vortex_bail!("Unsupported extension type \"{}\"", ext_dtype.id())
+                    // See if any dynamically registered kernels can handle it.
+                    for converter in inventory::iter::<DTypeConversionRef> {
+                        if converter.0.can_convert_to_arrow(ext_dtype.as_ref()) {
+                            return converter.0.to_arrow(ext_dtype.as_ref());
+                        }
+                    }
+                    vortex_bail!(
+                        "No registered converter for extension type \"{}\"",
+                        ext_dtype.id()
+                    )
                 }
             }
         })
     }
 }
 
+/// Convert a Vortex logical type into an Arrow physical type.
+///
+/// This function will perform lookups for plugins that are available at link time which implement
+/// the [`DTypeConversion`] trait.
+///
+/// See [`DTypeConversionRef`] documentation for more information.
+pub fn try_to_arrow(dtype: &DType) -> VortexResult<Field> {
+    dtype.to_arrow_field()
+}
+
+/// Type-erased pointer to a [`DTypeConversion`] implementation.
+pub struct DTypeConversionRef(ArcRef<dyn DTypeConversion>);
```

Comment:
I think this is the big question right. About whether extension DTypes themselves can extend behaviour, or whether they are just defined as metadata bytes and can extend behaviour through encodings.

If we do allow them to extend behaviour, I think I would want a `dyn ExtDType`, to avoid the marshalling of metadata bytes all over the place. Same reason we have `dyn Array` instead of the old `ArrayData` struct.